diff --git a/video_filter.module b/video_filter.module
index b21ad7b..f6aad38 100644
--- a/video_filter.module
+++ b/video_filter.module
@@ -250,6 +250,16 @@ function _video_filter_process($text, $filter, $format, $langcode, $cache, $cach
         }
         $video['height'] += $control_bar_height;
 
+        // Resize to fit within width and height repecting aspect ratio.
+        if ($ratio) {
+          $scale_factor = min(array(
+            ($video['height'] - $control_bar_height),
+            $video['width'] / $ratio,
+          ));
+          $video['height'] = round($scale_factor + $control_bar_height);
+          $video['width'] = round($scale_factor * $ratio);
+        }
+
         $video['autoplay'] = (bool) $video['autoplay'];
         $video['align'] = (isset($video['align']) && in_array($video['align'], array(
           'left',
