diff --git a/includes/bootstrap.inc b/includes/bootstrap.inc
index 728e4ec..b9f9d7a 100644
--- a/includes/bootstrap.inc
+++ b/includes/bootstrap.inc
@@ -912,7 +912,7 @@ function variable_initialize($conf = array()) {
     }
     else {
       // Proceed with variable rebuild.
-      $variables = array_map('unserialize', db_query('SELECT name, value FROM {variable}')->fetchAllKeyed());
+      $variables = array_map('unserialize', db_query('SELECT name, value FROM {variable} WHERE value <> :null', array(':null' => 'N;'))->fetchAllKeyed());
       cache_set('variables', $variables, 'cache_bootstrap');
       lock_release($name);
     }
@@ -989,14 +989,7 @@ function variable_set($name, $value) {
  * @see variable_set()
  */
 function variable_del($name) {
-  global $conf;
-
-  db_delete('variable')
-    ->condition('name', $name)
-    ->execute();
-  cache_clear_all('variables', 'cache_bootstrap');
-
-  unset($conf[$name]);
+  variable_set($name, NULL);
 }
 
 /**
